linked_list/linked_list.py

```diff
@@ -9,71 +9,142 @@ def __init__(self, value, next_node = None):
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        new_node = Node(value)
+        current = self.head
+        new_node.next = current #assigning the pointer in node
+        self.head = new_node
+        
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        current = self.head
+        while current != None:
+            if current.value == value:
+                return True
+            if current.value != value:
+                current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        current = self.head
+        total = 0
+        while current != None:
+            total += 1
+            current =current.next
+        return total
+        
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current = self.head
+        count = 0
+        while current != None:
+            if count > index:
+                return None
+            if count == index:
+                return current.value
+            else:
+                count += 1
+                current = current.next
+
+            
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current = self.head
+        while current != None:
+            if current.next == None:
+                return current.value
+            current = current.next
+
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        current = self.head
+        new_node = Node(value)
+        while current != None:
+            if current.next == None:
+                current.next = new_node
+                return
+            current = current.next
+        if current == None:
+            self.head = new_node
+            
+
+            
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):

```

Comment:
✨ Might consider initializing `max_value` to `None` or a maximum negative integer to allow this function to work with negative values.